### PR TITLE
Create replay detector for each SRTCP sender SSRC

### LINKED ISSUE #60: ReplayDetector is shared among multiple SSRCs
`srtp.Context` is shared among multiple SSRCs, but roll over counter and replay detector must not be shared.
Shared ROC causes decryption failure when the difference of actual ROCs is more than 1.
(Packet rate of each stream can be different.)
And, replay detector causes false positive detection.

ref: https://github.com/pion/webrtc/pull/1095#issuecomment-603169345

### LINKED ISSUE #68: Replay detector marking PLI packets from Firefox as duplicates
### Your environment.
  - Version: aa6cbe8 or 1.3.1
  - Browser: Firefox
 
### Other Information

[pcap-logs.zip](https://github.com/pion/srtp/files/4560774/pcap-logs.zip) contains:

1. `firefox-rtcp-from-log.pcap`: obtained using [this method](https://blog.mozilla.org/webrtc/debugging-encrypted-rtp-is-more-fun-than-it-used-to-be/)
2. `firefox-rtcp-received-tcpdump-from-srtp.pcap`: actual received packets, failed example, correlates to the first file
3. `chrome-rtcp-received-tcpdump-from-srtp.pcap`: working example, but taken with the replay detection commented out. But Chrome does _not_ fail even with replay detector turned on.

The last two files contain duplicate (sent and received) packets, they can be filtered out using the `frame.p2p_dir == sent` filter. These logs were obtained by modifying `session_srtcp.go`:

```diff
$ git diff
diff --git a/session_srtcp.go b/session_srtcp.go
index 3df39c5..2ecfe87 100644
--- a/session_srtcp.go
+++ b/session_srtcp.go
@@ -11,6 +11,17 @@ import (

 const defaultSessionSRTCPReplayProtectionWindow = 64

+var udpConn *net.UDPConn
+
+func init() {
+       fmt.Println("-- srtp udp --")
+       var err error
+       udpConn, err = net.ListenUDP("udp", &net.UDPAddr{IP: net.IP{127, 0, 0, 1}, Port: 5678})
+       if err != nil {
+               panic(err)
+       }
+}
+
 // SessionSRTCP implements io.ReadWriteCloser and provides a bi-directional SRTCP session
 // SRTCP itself does not have a design like this, but it is common in most applications
 // for local/remote to each have their own keying material. This provides those patterns
@@ -145,9 +156,15 @@ func destinationSSRC(pkts []rtcp.Packet) []uint32 {
 func (s *SessionSRTCP) decrypt(buf []byte) error {
        decrypted, err := s.remoteContext.DecryptRTCP(buf, buf, nil)
        if err != nil {
+               fmt.Println("Error decrypting RTCP:", err)
                return err
        }

+       _, err = udpConn.WriteTo(decrypted, &net.UDPAddr{IP: net.IP{127, 0, 0, 1}, Port: 5000})
+       if err != nil {
+               fmt.Println("Error writing to UDP 127.0.0.1:5000", err)
+       }
+
        pkt, err := rtcp.Unmarshal(decrypted)
        if err != nil {
                return err
```

### What did you do?

Error is reproducible using https://github.com/peer-calls/firefox-pli-bug

### What did you expect?

I expected to receive a PLI packet for 2nd track in Firefox.

### What happened?

The PLI packets were filtered out because of this line: https://github.com/pion/srtp/blob/aa6cbe8d27930b16d09ed9258f02a907f47beae1/srtcp.go#L32

Possibly related to #60.

### PR DESCRIPTION
**Pull request by at-wat**
Replay detector was shared among all SSRCs in same context.

### TODO
- Add test: done
- Fix encoder as well https://github.com/pion/srtp/issues/70

### Reference issue
Fixes #68

### CODE REVIEW

**Comment by jeremija** [comment]
@at-wat I confirmed this fixes the Firefox PLI bug, thanks so much!

**Comment by jeremija** [comment]
Do you know where does the SenderReport come from? I couldn't find anything about this packet in Firefox logs...

**Comment by at-wat** [comment]
Added a test to reproduce the bug.

**Comment by at-wat** [comment]
> Do you know where does the SenderReport packet come from?

I'm not sure, but I guess it's from browser?
Sender address and port can be found by packet dump.

File: context.go [hunk 1]
```diff
@@ -39,38 +39,45 @@ const (
 	srtcpIndexSize = 4
 )
 
-// Encode/Decode state for a single SSRC
-type ssrcState struct {
+// Encode/Decode state for a single SRTP SSRC
```
Comment (Sean-Der) [inline context.go:42]: `Encrypt/Decrypt state...`

Comment (Sean-Der) [reply context.go:42]: I am gonna fix this myself, don't worry about it. Then will merge :)

**Pull request merged**